pkg/bucketaccess/bucketaccess_controller.go

```diff
@@ -0,0 +1,248 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucketaccess
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	kubeerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	kube "k8s.io/client-go/kubernetes"
+	kubecorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	buckets "sigs.k8s.io/container-object-storage-interface-api/clientset"
+	bucketapi "sigs.k8s.io/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// BucketAccessListener manages Bucket objects
+type BucketAccessListener struct {
+	provisionerClient cosi.ProvisionerClient
+	provisionerName   string
+
+	kubeClient   kube.Interface
+	bucketClient buckets.Interface
+	kubeVersion  *utilversion.Version
+
+	namespace string
+}
+
+// NewBucketAccessListener returns a resource handler for BucketAccess objects
+func NewBucketAccessListener(provisionerName string, client cosi.ProvisionerClient) (*BucketAccessListener, error) {
+	ns := os.Getenv("POD_NAMESPACE")
+	if ns == "" {
+		return nil, errors.New("POD_NAMESPACE env var cannot be empty")
+	}
+
+	return &BucketAccessListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+		namespace:         ns,
+	}, nil
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
+// Return values
+//    nil - BucketAccess successfully granted
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (bal *BucketAccessListener) Add(ctx context.Context, bucketAccess *v1alpha1.BucketAccess) error {
+	bucketName := bucketAccess.Spec.BucketName
+
+	klog.V(3).InfoS("Add BucketAccess",
+		"name", bucketAccess.Name,
+		"bucket", bucketName,
+	)
+
+	if bucketAccess.Spec.MintedSecretName != "" {
+		klog.V(5).InfoS("AccessAlreadyGranted",
+			"bucketAccess", bucketAccess.Name,
+			"bucket", bucketName,
+		)
+		return nil
+	}
+
+	bucket, err := bal.Buckets().Get(ctx, bucketName, metav1.GetOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to fetch bucket", "bucket", bucketName)
+		return errors.Wrap(err, "Failed to fetch bucket")
+	}
+
+	if !strings.EqualFold(bucket.Spec.Provisioner, bal.provisionerName) {
```

Comment:
can this check be moved to the beginning of the func?

Reply:
It cannot. Because provisioner is inferred from the bucket